orchestrator/test_runner/src/erc_721_happy_path.rs

```diff
@@ -0,0 +1,638 @@
+            None,
+        )
+        .await;
+    }
+
+    let event_nonce = get_event_nonce_safe(gravity_address, web30, *MINER_ADDRESS)
+        .await
+        .unwrap();
+
+    // We are going to submit a duplicate tx with nonce 1
+    // This had better not increase the balance again
+    // this test may have false positives if the timeout is not
+    // long enough. TODO check for an error on the cosmos send response
+    submit_duplicate_erc20_send(
+        event_nonce, // Duplicate the current nonce
+        contact,
+        erc20_address,
+        1u64.into(),
+        user_keys.cosmos_address,
+        &keys,
+    )
+    .await;
+
+    // we test a batch by sending a transaction
+    test_batch(
+        contact,
+        &mut grpc_client,
+        web30,
+        user_keys.eth_address,
+        gravity_address,
+        keys[0].validator_key,
+        user_keys.cosmos_key,
+        erc20_address,
+    )
+    .await;
+}
+
+// Iterates each attestation known by the grpc endpoint by calling `get_attestations()`
+// Executes the input closure `f` against each attestation's decoded claim
+// This is useful for testing that certain attestations exist in the oracle,
+// see the consumers of this function (check_valset_update_attestation(),
+// check_send_to_cosmos_attestation(), etc.) for examples of usage
+//
+// `F` is the type of the closure, a state mutating function which may be called multiple times
+// `F` functions take a single parameter of type `T`, which is some sort of `Message`
+// The type `T` is very important as it dictates how we decode the message
+pub async fn iterate_attestations<F: FnMut(T), T: Message + Default>(
+    grpc_client: &mut GravityQueryClient<Channel>,
+    f: &mut F,
+) {
+    let attestations = get_attestations(grpc_client, None)
+        .await
+        .expect("Something happened while getting attestations after delegating to validator");
+    for (i, att) in attestations.into_iter().enumerate() {
+        let claim = att.clone().claim;
+        trace!("Processing attestation {}", i);
+        if claim.is_none() {
+            trace!("Attestation returned with no claim: {:?}", att);
+            continue;
+        }
+        let claim = claim.unwrap();
+        let mut buf = BytesMut::with_capacity(claim.value.len());
+        buf.extend_from_slice(&claim.value);
+
+        // Here we use the `T` type to decode whatever type of message this attestation holds
+        // for use in the `f` function
+        let decoded = T::decode(buf);
+
+        // Decoding errors indicate there is some other attestation we don't care about
+        if decoded.is_err() {
+            debug!(
+                "Found an attestation which is not a {}: {:?}",
+                type_name::<T>(),
+                att,
+            );
+            continue;
+        }
+        let decoded = decoded.unwrap();
+        f(decoded);
+    }
+}
+
+pub async fn wait_for_nonzero_valset(web30: &Web3, gravity_address: EthAddress) {
```

Comment:
not needed

Reply:
Removed as we are not checking attestations and functions called by it